yt_dlp/extractor/arte.py

```diff
@@ -5,195 +5,220 @@
+                        m.group('stlang') == lc,  # if subtitles are present, we prefer them in the requested language
+                        not m.group('st'),        # but we prefer no subtitles otherwise
+                        not m.group('stm'),       # and we prefer not the hard-of-hearing subtitles if there are subtitles
+                    )))
+                )
+
+            # XXX: probably not worth warning about if no match
+
+            if stream['protocol'].startswith('HLS'):
+                fmts, subs = self._extract_m3u8_formats_and_subtitles(
+                    stream['url'], video_id=video_id, ext='mp4',
+                    m3u8_id=stream_version_code,
+                )
+
+                for fmt in fmts:
+                    fmt.update({
+                        'format_note': f'{stream_version.get("label", "unknown")} [{stream_version.get("shortLabel", "?")}]',
                         'language_preference': lang_pref,
-                        'format_note': format_note,
                     })
-                formats.extend(m3u8_formats)
+
+                formats.extend(fmts)
+                self._merge_subtitles(subs, target=subtitles)
                 continue
 
-            format = {
-                'format_id': format_id,
-                'language_preference': lang_pref,
-                'format_note': format_note,
-                'width': int_or_none(f.get('width')),
-                'height': int_or_none(f.get('height')),
-                'tbr': int_or_none(f.get('bitrate')),
-                'quality': qfunc(f.get('quality')),
-            }
-
-            if media_type == 'rtmp':
-                format['url'] = f['streamer']
-                format['play_path'] = 'mp4:' + f['url']
-                format['ext'] = 'flv'
-            else:
-                format['url'] = f['url']
-
-            formats.append(format)
-
-        # For this extractor, quality only represents the relative quality
-        # with respect to other formats with the same resolution
-        self._sort_formats(formats, ('res', 'quality'))
+            if stream['protocol'] in ('HTTPS', 'RTMP'):
+                formats.append({
+                    'format_id': f'{stream["protocol"]}-{stream_version_code}',
+                    'url': stream['url'],
+                    'format_note': f'{stream_version.get("label", "unknown")} [{stream_version.get("shortLabel", "?")}]',
+                    'language_preference': lang_pref,
+                    # 'ext': 'mp4',  # XXX: may or may not be necessary, at least for HTTPS
+                })
+                continue
+
+            self.report_warning(
+                f'Skipping stream with unknown protocol {stream["protocol"]}')
+
+            # XXX: chapters from stream['segments']?
+            # the JS also apparently looks for chapters in
+            # config['data']['attributes']['chapters'], but
+            # I am yet to find a video having those
```

Comment:
I am trying to keep line length… in line, so to speak. Some of it is because I find long single-line comments harder to read myself. Some of it is habit from style standards of other projects, though I realise yt-dlp is not particularly strict about this. Some of it is because my screen is relatively small and I sometimes drag my editor to the side so that it only take half of the screen, and I like being able to see the whole code in that state without enabling line-wrapping (which makes the structure of the code marginally less discernible).